### PR TITLE
[feature] support jaeger telemetry

### PR DESCRIPTION
**Pull request by 0xcb9ff9**
# Description

Logs are difficult to track complex logic, e.g. connection management.
This PR uses Jaeger to provides a basic debugging environment.

Add flag:

- `jaeger`  enable the jaeger telemetry
- `jaeger-address`  jaeger thrift protocol endpoint

# Changes include

- [x] New feature (non-breaking change that adds functionality)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels

## Testing

- [x] I have tested this code with the official test suite
- [x] I have tested this code manually

### Manual tests
1.  Install local jaeger server [docs](https://www.jaegertracing.io/docs/1.42/getting-started/#all-in-one)
2.  Run `dogechain` node, but add `jaeger` flag `--jaeger` `--jaeger-address "http://127.0.0.1:14268/api/traces"`
       example: `dogechain server --chain ./genesis.json --jaeger --jaeger-address "http://127.0.0.1:14268/api/traces"`

### CODE REVIEW

File: command/helper/helper.go [hunk 1]
```diff
@@ -259,6 +259,29 @@ func GetPprofFlag(cmd *cobra.Command) bool {
...
+	cmd.PersistentFlags().String(
+		command.JaegerAddressFlag,
+		fmt.Sprintf("http://%s:%d/api/traces", LocalHostBinding, server.DefaultJaegerPort),
+		"jaeger thrift protocol (HTTP POST to /api/traces)",
+	)
+}
+
+// GetJaegerFlag extracts the telemetry flag
+func GetJaegerFlag(cmd *cobra.Command) (bool, string) {
+	v, _ := cmd.Flags().GetBool(command.JaegerFlag)
```
Comment (DarianShawn) [inline command/helper/helper.go]: Not handle error?

File: helper/telemetry/type_convert.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
+package telemetry
+
+import (
+	"go.opentelemetry.io/otel/attribute"
+
+	"fmt"
+)
+
+func convertTypeToAttribute(value interface{}) attribute.Value {
+	switch v := value.(type) {
```
Comment (DarianShawn) [inline helper/telemetry/type_convert.go:10]: No pointer value?

File: network/server.go [hunk 3]
```diff
@@ -602,6 +576,16 @@ func (s *DefaultServer) Peers() []*PeerConnInfo {
...
 }
 
+// HasFreeConnectionSlot checks if there are free connection slots in the specified direction [Thread safe]
+func (s *DefaultServer) HasFreeConnectionSlot(direction network.Direction) bool {
+	return s.connectionCounts.HasFreeConnectionSlot(direction)
+}
+
+// GetTracer returns the tracer instance
+func (s *DefaultServer) GetTracer() telemetry.Tracer {
+	return s.tracer
```
Comment (DarianShawn) [inline network/server.go:586]: Is it thread safe to use the same tracer?

File: network/peer_connect.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+// addProtocolClient adds a protocol stream
+func (pci *PeerConnInfo) addProtocolClient(protocol string, stream client.GrpcClientCloser) {
+	pci.protocolClient[protocol] = stream
+}
+
+// cleanProtocolStreams clean and closes all protocol stream
+func (pci *PeerConnInfo) cleanProtocolStreams(ctx context.Context, trace telemetry.Tracer) []error {
+	errs := []error{}
+	span := trace.StartWithContext(ctx, "PeerConnInfo.cleanProtocolStreams")
```
Comment (DarianShawn) [inline network/peer_connect.go:59]: Inner span doesn't end on return?

File: network/server.go [hunk 5]
```diff
@@ -704,7 +708,14 @@ func (s *DefaultServer) ForgetPeer(peer peer.ID, reason string) {
 
 // DisconnectFromPeer disconnects the networking server from the specified peer
 func (s *DefaultServer) DisconnectFromPeer(peerID peer.ID, reason string) {
```
Comment (DarianShawn) [inline network/server.go:710]: I think it sounds more reasonable if it accept `context.Context` parameter.
All the other external methods in this module should use context to preserve its span tree, too.

**Review by DarianShawn — COMMENTED** [review]
I love this great idea with bringing an open tracer standard into project to make tracking clearer.

Some more suggestions here though:

* Use the same context as much as possible.
  * To exactly unveil our complex calling hierarchy.
* Use server option mode for more and more flags?
  * To better tidy and expand our feature flags.
  * Don't do it in current PR any way.

File: helper/telemetry/type_convert.go (hunk 2 above)
Comment (0xcb9ff9) [reply helper/telemetry/type_convert.go:10]: attribute only support base value types in design, pointer value not thread safe

File: network/server.go (hunk 3 above)
Comment (0xcb9ff9) [reply network/server.go:586]: `telemetry.Tracer` is [Tracer](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#TracerProvider.Tracer)  wrapping, is thread safe

File: network/peer_connect.go (hunk 4 above)
Comment (0xcb9ff9) [reply network/peer_connect.go:59]: fix it [c783e29](https://github.com/dogechain-lab/dogechain/pull/317/commits/c783e29faa17eff153ac70d74ef9a87830725ff9)

File: network/server.go (hunk 5 above)
Comment (0xcb9ff9) [reply network/server.go:710]: this is big change, need split

File: helper/telemetry/type_convert.go (hunk 2 above)
Comment (DarianShawn) [reply helper/telemetry/type_convert.go:10]: Got it.

**Review by DarianShawn — APPROVED** [review]
LGTM

**Pull request merged**